rfcs/033-api-internal-model-versioning/README.md

```diff
@@ -0,0 +1,78 @@
+
+- We need integration tests that tell us if a build of the API is compatible with the currently available index and we
+  can't guarantee that what is in master is always deployable and that integration tests should always pass.
+
+## Proposed solution
+
+The proposed solution is:
+- for every build in master, release the internal model as an artifact with a new version
+  and publish it to our s3 respository in a similar way as to how we release artifacts in the scala-libs repo.
+
+- all catalogue pipeline services that depend on the internal model keep the dependency on the local module
+  instead of the published artifact. When we work a something that require model changes, we work on our dev branch
+  on model and pipeline services at the same time and once the branch is merged, a new version of the internal model
+  is published.
+
+- the catalogue API depends on the internal model artifact published in s3 instead of the internal model module.
+  We update the version the API depends on in master only when we have an index compatible with that version. Any change
+  in the API that does not require model changes can be merged in master and deployed without pointing the API to a new index.
+
+- If we need to make changes both on the pipeline and on the API, we need to make that in three stages:
+  - change the pipeline
+  - reindex
+  - change the API
+
+## Advantages
+
+This means that:
+- changes to the API that don't need internal model changes can be deployed independently of reindexes
+  and independently of any internal model change.
+
+- If there are problems in the pipeline causing reindex delays, we can still deploy changes to the API,
+  provided that we don't upgrade the version of the internal model dependency.
+
+- We can fix bugs in the API by releasing a new version of the API if we don't need to upgrade model version.
+
+- We can adopt a policy where we say integration tests should always pass and the API should always be deployable
+  from master to production.
+
+- We can split the API into its own repo which could allow us to think substantially differently about the API than
+  the pipeline and have a different build.
+
+
+## Potential problems with the proposed approach
+
+- It's difficult to know from an index what version of the internal model was used to populate it. Ideally we might want to include
+  the model version in the index name but because the pipeline services, including the ingestor, depend on the local internal_model,
+  they don't know the version, so it's not straightforward to do. This means that we could run into issues if we update the index tha API
+  is pointing to without updating the model dependency and viceversa.
+
+- If we need to make changes to the API that require internal model changes, we need to wait on a new release of the internal model, before
+  we can start working on it.
```

Comment:
I don't think this is necessarily an issue – if we do this, I imagine we'd do pipeline and API patches separately.

The pipeline patch gets merged immediately, gets deployed and we reindex. That patch includes the internal model change, and when it gets merged we publish a new version.

That then allows you to write the API patch separately, but you don't merge and deploy it until the reindex is done and you have an index ready to go. That PR might be open for a while, but it's not merged until you're almost ready to deploy it.

Reply:
Yeah, I don't think it's an issue but it is a difference compared to what we can do right now. For example, we might do all the pipeline related changes, then move on to the api changes and realise we forgot something on the pipeline side. That means reopening a PR, a new release etc. I think it probably doesn't add that much in terms of time lost and anyway the benefits outweigh the downsides but I thought it was woth mentioning it. 

Perhaps it could be worded better though